src/expr/attribute_internals.h

```diff
@@ -145,16 +149,372 @@ inline uint64_t GetBitSet(uint64_t bit)
+    }
+
+    // range insert
+    template <typename Iter>
+    void insert(Iter beg, Iter end)
+    {
+      for (Iter it = beg; it != end; ++it)
+      {
+        iterator found_it =
+            std::lower_bound(d_contents.begin(),
+                             d_contents.end(),
+                             it->first,
+                             [](const value_type& a, const value_type& b) {
+                               return a.first < b.first;
+                             });
+        if ((found_it != d_contents.end()) && (it->first == found_it->first))
+        {
+          // this key is already present in the map. replace it:
+          found_it->second = it->second;
+        }
+        else
+        {
+          d_contents.insert(found_it, *it);
+        }
+      }
+    }
+
+   private:
+    Container d_contents;
+  };
+
+  template <typename Parent, typename L1It, typename L2It>
+  class Iterator
+  {
+    // access for AttrHash to implement erase(Iterator)
+    using NonConstParent = typename std::remove_const_t<Parent>;
+    using NonConstIterator = typename NonConstParent::iterator;
+    friend NonConstIterator Parent::erase(NonConstIterator);
+
+   public:
+    // requirements for ForwardIterator
+    using iterator_category = std::forward_iterator_tag;
+    using value_type = std::pair<std::pair<uint64_t, NodeValue*>, V>;
+    using reference = value_type;  // we don't supply a true reference
+    using pointer = value_type*;
+    using difference_type = std::ptrdiff_t;
+
+    // default constructor
+    Iterator() : d_atEnd{true} {}
+
+    Iterator(Parent* parent)
+        : d_parent{parent}, d_l1It(parent->d_storage.begin())
+    {
+      d_atEnd = (d_l1It == parent->d_storage.end());
+      if (!d_atEnd)
+      {
+        d_l2It = d_l1It->second.begin();
+        legalize();  // L2 map may be empty
+      }
+    }
+
+    // prerequisite: l1_it and l2_it are valid iterators
+    Iterator(Parent* parent, L1It l1_it, L2It l2_it)
+        : d_atEnd(l1_it == parent->d_storage.end()),
+          d_parent(parent),
+          d_l1It(l1_it),
+          d_l2It(l2_it)
+    {
+    }
+
+    // increment
+    Iterator& operator++()  // pre
+    {
+      increment();
+      return *this;
+    }
+
+    Iterator operator++(int)  // post
+    {
+      Iterator tmp = *this;
+      increment();
+      return tmp;
+    }
+
+    // dereference
+    value_type operator*() const
+    {
+      return std::make_pair(std::make_pair(d_l2It->first, d_l1It->first),
+                            d_l2It->second);
+    }
+
+    // comparison
+    bool operator==(Iterator const& other) const
+    {
+      return (d_atEnd && other.d_atEnd)
+             || (!d_atEnd && !other.d_atEnd && (d_l1It == other.d_l1It)
+                 && (d_l2It == other.d_l2It));
+    }
+    bool operator!=(Iterator const& other) const { return !(*this == other); }
+
+   private:
+    void increment()
+    {
+      ++d_l2It;
+
+      legalize();
+    }
+
+    void legalize()
```

Comment:
It would be great if you could document what `legalize()` does and when it should be called.